brave/src/main/java/brave/propagation/StrictScopeDecorator.java

```diff
@@ -30,38 +35,66 @@
  *                  ).build();
  * }</pre>
  */
-public final class StrictScopeDecorator implements ScopeDecorator {
-  public static ScopeDecorator create() {
+// Closeable so things like Spring will automatically execute it on shutdown and expose leaks!
+public final class StrictScopeDecorator implements ScopeDecorator, Closeable {
+  public static StrictScopeDecorator create() {
     return new StrictScopeDecorator();
   }
 
-  /** Identifies problems by throwing assertion errors when a scope is closed on a different thread. */
-  @Override public Scope decorateScope(@Nullable TraceContext currentSpan, Scope scope) {
-    return new StrictScope(scope, new Error(String.format("Thread %s opened scope for %s here:",
-      Thread.currentThread().getName(), currentSpan)));
+  final Queue<CallerStackTrace> currentCallers = new ConcurrentLinkedQueue<>();
+
+  /**
+   * Identifies problems by throwing {@link IllegalStateException} when a scope is closed on a
+   * different thread.
+   */
+  @Override public Scope decorateScope(@Nullable TraceContext context, Scope scope) {
+    return new StrictScope(scope, new CallerStackTrace(context), currentCallers);
+  }
+
+  /**
+   * @throws IllegalStateException if any scopes were left unclosed.
+   * @since 5.11
+   */
+  @Override public void close() {
+    for (CallerStackTrace caller : currentCallers) {
+      throw new IllegalStateException(
+        "Thread [" + caller.threadName + "] opened scope, but never closed it", caller);
+    }
   }
 
   static final class StrictScope implements Scope {
     final Scope delegate;
-    final Throwable caller;
-    final long threadId = Thread.currentThread().getId();
+    final Queue<CallerStackTrace> currentCallers;
+    final CallerStackTrace caller;
 
-    StrictScope(Scope delegate, Throwable caller) {
+    StrictScope(Scope delegate, CallerStackTrace caller, Queue<CallerStackTrace> currentCallers) {
       this.delegate = delegate;
+      this.currentCallers = currentCallers;
       this.caller = caller;
+      this.currentCallers.add(caller);
     }
 
     @Override public void close() {
-      if (Thread.currentThread().getId() != threadId) {
-        throw new IllegalStateException(
-          "scope closed in a different thread: " + Thread.currentThread().getName(),
-          caller);
+      currentCallers.remove(caller);
```

Comment:
Probably don't care about performance of this but just in case, can consider using a `LinkedBlockingDeque` to just remove the last element instead of iterating to the end.

Reply:
I so wanted to, but that's Java 1.7 plus (animal sniffer found this)